src/main/java/uk/gov/companieshouse/api/accounts/service/impl/CompanyAccountServiceImpl.java

```diff
@@ -1,46 +1,93 @@
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
 import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.beans.factory.annotation.Qualifier;
-import org.springframework.data.mongodb.repository.MongoRepository;
 import org.springframework.stereotype.Service;
-import uk.gov.companieshouse.api.accounts.Kind;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+import uk.gov.companieshouse.GenerateEtagUtil;
+import uk.gov.companieshouse.api.accounts.AttributeName;
+import uk.gov.companieshouse.api.accounts.LinkType;
 import uk.gov.companieshouse.api.accounts.model.entity.CompanyAccountEntity;
 import uk.gov.companieshouse.api.accounts.model.rest.CompanyAccount;
+import uk.gov.companieshouse.api.accounts.repository.CompanyAccountRepository;
 import uk.gov.companieshouse.api.accounts.service.CompanyAccountService;
-import uk.gov.companieshouse.api.accounts.transformer.GenericTransformer;
+import uk.gov.companieshouse.api.accounts.transaction.Transaction;
+import uk.gov.companieshouse.api.accounts.transaction.TransactionManager;
+import uk.gov.companieshouse.api.accounts.transformer.CompanyAccountTransformer;
 
 @Service
-public class CompanyAccountServiceImpl extends
-        AbstractServiceImpl<CompanyAccount, CompanyAccountEntity> implements CompanyAccountService {
+public class CompanyAccountServiceImpl implements CompanyAccountService {
 
     @Autowired
-    public CompanyAccountServiceImpl(
-            @Qualifier("companyAccountRepository") MongoRepository<CompanyAccountEntity, String> mongoRepository,
-            @Qualifier("companyAccountTransformer") GenericTransformer<CompanyAccount, CompanyAccountEntity> transformer) {
-        super(mongoRepository, transformer);
+    private TransactionManager transactionManager;
+
+    @Autowired
+    private CompanyAccountRepository companyAccountRepository;
+
+    @Autowired
+    private CompanyAccountTransformer companyAccountTransformer;
+
+    /**
+     * {@inheritDoc}
+     */
+    public CompanyAccount createCompanyAccount(CompanyAccount companyAccount) {
+        HttpServletRequest request = getRequestFromContext();
+        Transaction transaction = getTransactionFromSession(request);
+
+        String id = generateID();
+        String companyAccountLink = createSelfLink(transaction, id);
+        addKind(companyAccount);
+        addEtag(companyAccount);
+        addLinks(companyAccount, companyAccountLink);
+
+        CompanyAccountEntity companyAccountEntity = companyAccountTransformer.transform(companyAccount);
+
+        companyAccountEntity.setId(id);
+        companyAccountRepository.insert(companyAccountEntity);
+
+        transactionManager.updateTransaction(transaction.getId(),
+                request.getHeader("X-Request-Id"),
+                companyAccountLink);
+
+        return companyAccount;
```

Comment:
we should not be returning the 'rest' model that we passed into this method, we need the ability to return multiple things here i.e. the response code and body.

Brain dump from several of us:

The easiest way to do this is to create the ResponseEntity with the correct status code and body within the service and return that - this means any errors from the db insert or transaction patch can explicitly set their response codes rather than hoping the global exception handler gets it right.
As the Controller should be dealing with the response codes and body to return we could create a utility object or something similar that can have multiple fields that get mapped to the response code/body by the controller.

Another way to do this would be to throw different exceptions for each case and catch those in the controller which will the decide the response code and body to return if necessary however that is not very nice or explicit and require more logic in the controller.

Reply:
I like the second method more again best practice in attempting to maintain a clear separation of concerns. You have to think that each Service may be used by multiple Controllers and attempting to code for each Controllers needs in the service layer will add to much complexity.

Reply:
I agree, just means that the controllers will have more logic in them - which is fine.
If this is the way we go then the controller needs to explicitly handle each distinct sunny day and failure case e.g. in this specific function it should handle :
- successful write of company-accounts data and successful patch of transaction
- failed write of company-accounts to mongo
- failed transaction patch
Each of these should be explicitly caught/checked in the controller, maybe requiring this service method to be split up so the controller makes multiple calls into the service,possibly 1 for decorating the REST object with kind, links, etag, another for transformation to DB model and writing to mongo and another for patching transaction

Reply:
The Controllers will have to interpret anything that is returned from the service layer or below as is their job, it follows the throw early, catch late paradigm by design. I agree with outlined functionality detailed above but the error handling isn't part of this PR, we need to address that when we have more of the system in place this is just to complete the logic to get a happy path in place for future work to build on.

Reply:
Error handling is part of every PR and should not have to be back ported into code that is already in develop. 
It should be included here even if it's a basic try-catch with a generic exception that will be replaced with more granular exception handling per case later.

Reply:
I will throw generic errors at the time being, and then we can improve those error types in the future. At least there's something there to cover the scenarios.